Create errors lesson (#61)
Ada-Developers-Academy/ada-precourse-v2: 01-intro-to-python/08-errors.md

    print(f"Hello {preson}!")
NameError: name 'preson' is not defined
```

It looks like we've got a bug in our program! This error output is quite a mouthful, so let's split it into parts to see what it's telling us.

## Dissecting an error

Errors in Python often have two parts - a stack trace and an error message. The stack trace will tell us where the error happened, and the error message will tell us what went wrong. We will discuss each part.

### Stack Trace

| Vocab       | Definition                                                                                                                                            | Synonyms  | How to Use in a Sentence                                                                                                                                                           |
| ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Stack Trace | A report of the functions that were called certain point in time during the execution of a program. | Traceback | "When debugging that error, we should check the stack trace to see what line of code caused it," "The stack trace is very long, but we can see what test called this function here." |

Reviewer: I think a couple words are missing in the Stack Trace definition:

```suggestion
| Stack Trace | A report of the functions that were called at a certain point in time during the execution of a program. | Traceback | "When debugging that error, we should check the stack trace to see what line of code caused it," "The stack trace is very long, but we can see what test called this function here." |
```

> | Stack Trace | A report of the functions that were called at a certain point in time during the execution of a program. | Traceback | "When debugging that error, we should check the stack trace to see what line of code caused it," "The stack trace is very long, but we can see what test called this function here." |

Author: Sounds like a good idea. Where's the best place to note a task like this?

> | Stack Trace | A report of the functions that were called at a certain point in time during the execution of a program. | Traceback | "When debugging that error, we should check the stack trace to see what line of code caused it," "The stack trace is very long, but we can see what test called this function here." |

Reviewer 2: Maybe either an issue in the core-unit-1 repo (though we don't really have a strong culture of noticing those), or in the [c16 feedback doc](https://docs.google.com/document/d/1yhl8krK_flu7s18GHWaJsvF1vwBHhp31ZxYKJbKyR4c/edit?usp=sharing), which might actually get a look-at when unit 1 is revisited.

Or go ahead and do it now? 😆

```
</details>

## Debugging tips

### The good stuff is usually at the bottom

Stack traces can get VERY long if there's lots of function calls. Usually, you don't need to look at the whole thing. Start by looking just at the error message, and the last part of the stack trace. This will tell you the error that happened, and the line where the error actually occurred.

### Search engines are your friend!

What if we looked at the error message and it wasn't clear to us what it was saying? We can search for the error message online to help us find out what's going on!

Our error message will often include some of our own variable names that might obscure the search. Try removing these before pasting the error message into the search engine. For example, instead of searching for:

```
NameError: name 'preson' is not defined
```

Try searching for:

```
NameError: name is not defined
```

Note that we removed `'preson'` because that variable name is specific to our code.

When I searched this error, I got [this excellent page](https://careerkarma.com/blog/python-nameerror-name-is-not-defined/) that described a `NameError` and possible causes/fixes.

### VS Code is your friend!

> VS Code will often help you find bugs by underlining possible parts of your code with issues. For example, this is what our example code looks like in VS Code:

Reviewer 2: - voicing to we/us
- the two close by examples sound redundant to me
```suggestion
VS Code will often help us find bugs by underlining parts of your code with possible issues. This is what our example code looks like in VS Code:
```

> ![vscode_underline](/01-intro-to-python/images/vscode-underlining.png)

Reviewer 2: Descriptive alt text
```suggestion
![The sample code used in this lesson as displayed in VS Code. There is a squiggly line underneath the misspelled preson variable.](/01-intro-to-python/images/vscode-underlining.png)
```

Note that there's a squiggly underline beneath our variable name with the typo! Super useful!

> To get even more information, you can hover your mouse over the underlined portion:

Reviewer 2: ```suggestion
To get even more information, we can hover the mouse over the underlined portion:
```

> ![vscode_hover](/01-intro-to-python/images/vscode-hover.png)

Reviewer 2: ```suggestion
![The sample code used in this lesson as displayed in VS Code. A message box displays details about the underlined error: "preson" is not defined.](/01-intro-to-python/images/vscode-hover.png)
```

> We can see that it warns us that the variable is not defined. You may notice that the warning that VS Code gives when hovering is different than what we get when running the program. It's often useful to look at both to see if one is easier to understand than the other!

Reviewer 2: rephrase to avoid the you
```suggestion
We can see that it warns us that the variable is not defined. Sometimes the warning that VS Code gives when hovering is different than what we get when running the program. It's often useful to look at both to see if one is easier to understand than the other!
```

## Summary

> Reading error output is an essential skill for software engineers. In this lesson we learned how we can use the stack trace to see where an error happened and the error message to tell what went wrong. We now know that the useful part of the error output is often at the bottom. We also found that VS Code and search engines can be helpful in deciphering confusing error messages.

Reviewer 2: ```suggestion
Reading error output is an essential skill for software engineers. In this lesson we learned how we can use the stack trace to see where an error happened, and the error message to tell what went wrong. We now know that the useful part of the error output is often at the bottom. We also found that VS Code and search engines can be helpful in deciphering confusing error messages.
```

#### !end-explanation

### !end-challenge

### !challenge

* type: multiple-choice
* id: 938efb00-5764-4f6d-8a3c-38101e348416
* title: Reading Tracebacks

##### !question

Suppose we get an error output that looks like this:

```
Traceback (most recent call last):
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 16, in <module>
    a()
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 5, in a
    b()
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 11, in b
    something()
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 14, in something
    ada()
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 2, in ada
    lovelace()
  File "/Users/auberonlopez/programming/ada/scratch/number.py", line 8, in lovelace
    1 / 0
ZeroDivisionError: division by zero
```

What line does the `ZeroDivisionError` occur at in our code?
##### !end-question

##### !options

* 16
* 5
* 11
* 14
* 2
* 8

##### !end-options

##### !answer

* 8

##### !end-answer

#### !explanation

> Remember, the good stuff is often at the bottom! The last part of the stack trace tells us that the error happened on line 8. All the other parts of the stack trace are just telling us which functions were called on our way to line 8.

Reviewer 2: The curriculum has a strong preference to avoid "just" (sometimes a little overzealously, I think). But here I think it can safely be dropped.
```suggestion
Remember, the good stuff is often at the bottom! The last part of the stack trace tells us that the error happened on line 8. All the other parts of the stack trace are telling us which functions were called on our way to line 8.
```
